mongorestore/oplog.go

```diff
@@ -167,32 +203,136 @@ func (restore *MongoRestore) HandleNonTxnOp(oplogCtx *oplogContext, op db.Oplog)
-		if index.Key == nil {
-			return fmt.Errorf("failed to parse IndexDocument from createIndexes in %s, %v", collectionName, op)
-		}
+		ns := strings.Split(op.Namespace, ".")
+		dbName := ns[0]
 
-		indexes := []IndexDocument{index}
-		if restore.OutputOptions.ConvertLegacyIndexes {
-			indexes = restore.convertLegacyIndexes(indexes, op.Namespace)
-		}
+		switch cmdName {
+		case "commitIndexBuild":
+			// commitIndexBuild was introduced in 4.4, one "commitIndexBuild" command can contain several
+			// indexes, we need to convert the command to "createIndexes" command for each single index and apply
+			collectionName, indexes := extractIndexDocumentFromCommitIndexBuilds(op)
+			if indexes == nil {
+				return fmt.Errorf("failed to parse IndexDocument from commitIndexBuild in %s, %v", collectionName, op)
+			}
 
-		return restore.CreateIndexes(strings.Split(op.Namespace, ".")[0], collectionName, indexes, false)
+			if restore.OutputOptions.ConvertLegacyIndexes {
+				indexes = restore.convertLegacyIndexes(indexes, op.Namespace)
+			}
+
+			collName, ok := op.Object[0].Value.(string)
+			if !ok {
+				return fmt.Errorf("could not parse collection name from op: %v", op)
+			}
+
+			restore.indexCatalog.AddIndexes(dbName, collName, indexes)
+			return nil
+
+		case "createIndexes":
+			// server > 4.4 no longer supports applying createIndexes oplog, we need to convert the oplog to createIndexes command and execute it
+			collectionName, index := extractIndexDocumentFromCreateIndexes(op)
+			if index.Key == nil {
+				return fmt.Errorf("failed to parse IndexDocument from createIndexes in %s, %v", collectionName, op)
+			}
+
+			indexes := []*idx.IndexDocument{index}
+			if restore.OutputOptions.ConvertLegacyIndexes {
+				indexes = restore.convertLegacyIndexes(indexes, op.Namespace)
+			}
+
+			collName, ok := op.Object[0].Value.(string)
+			if !ok {
+				return fmt.Errorf("could not parse collection name from op: %v", op)
+			}
+
+			restore.indexCatalog.AddIndexes(dbName, collName, indexes)
+			return nil
+
+		case "dropDatabase":
+			restore.indexCatalog.DropDatabase(dbName)
+
+		case "drop":
+			collName, ok := op.Object[0].Value.(string)
+			if !ok {
+				return fmt.Errorf("could not parse collection name from op: %v", op)
+			}
+			restore.indexCatalog.DropCollection(dbName, collName)
+
+		case "applyOps":
+			rawOps, ok := op.Object[0].Value.(bson.A)
+			if !ok {
+				return fmt.Errorf("unknown format for applyOps: %#v", op.Object)
+			}
+
+			for _, rawOp := range rawOps {
+				bytesOp, err := bson.Marshal(rawOp)
+				if err != nil {
+					return fmt.Errorf("could not marshal applyOps operation: %v: %v", rawOp, err)
+				}
+				var nestedOp db.Oplog
+				err = bson.Unmarshal(bytesOp, &nestedOp)
+				if err != nil {
+					return fmt.Errorf("could not unmarshal applyOps command: %v: %v", rawOp, err)
+				}
+
+				restore.HandleOp(oplogCtx, nestedOp)
+				if err != nil {
+					return err
+				}
+			}
```

Comment:
Should we `return nil` here? I guess we don't want to run `restore.ApplyOps` on this op again, right?

Reply:
Yes, great catch!